MaterialDesignThemes.Wpf/SmartHint.cs

```diff
@@ -0,0 +1,179 @@
+            set { SetValue(HintOpacityProperty, value); }
+        }
+
+        #endregion
+
+        static SmartHint()
+        {
+            DefaultStyleKeyProperty.OverrideMetadata(typeof(SmartHint), new FrameworkPropertyMetadata(typeof(SmartHint)));
+        }
+
+        public SmartHint()
+        {
+            IsHitTestVisible = false;
+            HorizontalAlignment = HorizontalAlignment.Left;
+            VerticalAlignment = VerticalAlignment.Top;
+        }
+
+        private static void HintProxyPropertyChangedCallback(DependencyObject dependencyObject, DependencyPropertyChangedEventArgs dependencyPropertyChangedEventArgs)
+        {
+            var smartHint = dependencyObject as SmartHint;
+            if (smartHint == null) return;
+
+            var hintProxy = dependencyPropertyChangedEventArgs.OldValue as IHintProxy;
+
+            if (hintProxy != null)
+            {
+                hintProxy.IsVisibleChanged -= smartHint.OnHintProxyIsVisibleChanged;
+                hintProxy.TextChanged -= smartHint.OnHintProxyTextChanged;
+                hintProxy.Loaded -= smartHint.OnHintProxyTextChanged;
+                hintProxy.Dispose();
+            }
+
+            hintProxy = dependencyPropertyChangedEventArgs.NewValue as IHintProxy;
+            if (hintProxy != null)
+            {
+                hintProxy.IsVisibleChanged += smartHint.OnHintProxyIsVisibleChanged;
+                hintProxy.TextChanged += smartHint.OnHintProxyTextChanged;
+                hintProxy.Loaded += smartHint.OnHintProxyTextChanged;
+                smartHint.RefreshState(false);
+            }
+        }
+
+        protected virtual void OnHintProxyTextChanged(object sender, EventArgs e)
+        {
+            IsTextNullOrEmpty = String.IsNullOrEmpty(HintProxy.Text);
+
+            if (HintProxy.IsLoaded)
+            {
+                RefreshState(true);
+            }
+            else
+            {
+                HintProxy.Loaded += HintProxySetStateOnLoaded;
+            }
+        }
+
+        private void HintProxySetStateOnLoaded(object sender, EventArgs e)
+        {
+            RefreshState(false);
+            HintProxy.Loaded -= HintProxySetStateOnLoaded;
+        }
+
+        protected virtual void OnHintProxyIsVisibleChanged(object sender, EventArgs e)
+        {
+            RefreshState(false);
+        }
+
+        private void RefreshState(bool useTransitions)
+        {
+            if (HintProxy == null) return;
+            if (!HintProxy.IsVisible) return;
+
+            Dispatcher.BeginInvoke(new Action(() =>
+            {
+                var state = string.IsNullOrEmpty(HintProxy.Text)
+                    ? MaterialDesignStateTextEmptyName
+                    : MaterialDesignStateTextNotEmptyName;
+
+                VisualStateManager.GoToState(this, state, useTransitions);
+            }));
+        }
+    }
+
+    public interface IHintProxy : IDisposable
```

Comment:
Can we pop this into it's own file and add a couple of comments to the header...it's a pretty important interface.